Refactor Form Submit Actions

dongilbert opened the pull request:
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | N
| Related user documentation PR URL | N/A
| Related developer documentation PR URL | N/A
| Issues addressed (#s or URLs) | #8066 
| BC breaks? | Y
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
Form submit actions had supports for proper events added a while back. This PR leverages the form submit event by removing callback support, and refactoring the callbacks away from usage of MauticFactory.

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Ensure all form submit actions continue to function as expected.

#### List backwards compatibility breaks:
1. LeadEvents::FORM_SUBMIT_REMOVE_DO_NOT_CONTACT event is removed. Listen for the LeadEvents::ON_EXECUTE_SUBMIT_ACTION instead, and check the context of `lead.remove_do_not_contact`

escopecz commented:
@dongilbert could you rebase?

dongilbert commented:
Rebased to 3.x

dongilbert commented on app/bundles/NotificationBundle/EventListener/FormSubscriber.php:
```diff
@@ -177,8 +195,10 @@ public function onFormActionSend(SubmissionEvent $event)
-        if (200 !== $response->code) {
-            return $event->setResult(false);
+        if ($response->code !== 200) {
+            $event->setResult(false);
```
This subscriber is not registered in the config file, and is not used. We should delete it instead.

dongilbert commented on app/bundles/FormBundle/Tests/EventListener/FormSubscriberTest.php:
```diff
@@ -1,46 +0,0 @@
-use Mautic\FormBundle\Event\FormBuilderEvent;
-use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;
-
-class FormSubscriberTest extends WebTestCase
```
The test didn't test anything

ondrejsibl commented:
@dongilbert Very nice work, Don! One more thing I found now during testing, there are missing arguments for Mautic\EmailBundle\EventListener\FormSubscriber constructor in configuration.

dongilbert commented:
oops, added. Thanks for catching it

ondrejsibl commented:
I am now trying to test it, but I am getting errors. Maybe these errors are not caused by this task. 

```
TemplatingHelper.php line 48:
                                                                                                                  
Attempted to call an undefined method named "enterScope" of class "ContainerLtcoxmf\appProdProjectContainer".  
```
                         
<img width="950" alt="Snímek obrazovky 2019-12-10 v 14 48 53" src="https://user-images.githubusercontent.com/10339966/70535102-bcff2380-1b5c-11ea-9195-c3479d7bd22f.png">

dongilbert commented:
This is due to the Page Bundle not yet being updated in the 3.x branch. Nothing to do with this PR

ondrejsibl commented:
That's what I thought, but there are more problems in 3.x, so it is not testable for now.

dongilbert commented:
What specifically in regards to this PR is not testable?

ondrejsibl commented:
@dongilbert I can't prepare any testing data, I can't send emails. @escopecz was telling me, that we shouldn't test on that branch. But I can try to prepare the testing data on another branch, still not sure if I can test everything. 

I tried to submit a from inserted in a landing page, got this error:
 message = "Call to a member function getType() on null"
```
 *Exception*string = ""
 code = {int} 0
 file = "/Users/ondrejsibl/dev/apps/community/app/bundles/FormBundle/Event/SubmissionEvent.php"
 line = {int} 280
$exception = {Symfony\Component\Debug\Exception\FatalThrowableError} [8]
```

dongilbert commented:
Instead of embedding a form in the landing page (since landing pages are broken), you could just preview the form and test it that way

hluchas commented:
@dongilbert Is this still WIP as described in PR description? I'm confused little bit.

dongilbert commented:
@hluchas no, it's not WIP

hluchas commented:
@dongilbert Can you rebase 3.x branch? There are a lot of other problems with forms.

dongilbert commented:
It's already against the 3.x branch, and there are no merge conflicts - what issues are you having?

escopecz commented on app/bundles/LeadBundle/EventListener/FormSubscriber.php:
```diff
@@ -32,11 +37,34 @@ class FormSubscriber implements EventSubscriberInterface
     private $emailModel;
 
     /**
+<<<<<<< HEAD
```
@hluchas how did you miss this? @dongilbert can you resolve this? Also, removing several classes is not considered a BC break?

dongilbert replied:
Probably because it was in comments - I know that's how I missed it. 

I'll document the removed classes specifically.

dongilbert commented:
@escopecz I've updated https://docs.google.com/document/d/1lHNK8hHQtvyWlz9OtfAoLupDVb5-QunQjq-HAGlAsS4/edit# with the removed classes and fixed the commented out merge conflict

hluchas commented:
@dongilbert Conflict is resolved. You can merge I think.

escopecz commented:
It's not related to this PR, but I'm getting error

`PHP Warning - array_flip(): Can only flip STRING and INTEGER values! in app/bundles/FormBundle/Form/Type/FieldType.php:395`

when I try to add an email or a text field to the form. It's making it hard to test this PR.

escopecz commented:
Another issue. When I try to create a new form action "Push contact to integration"

`PHP Notice - Undefined variable: integrationHelper in app/bundles/PluginBundle/Form/Type/IntegrationsListType.php:85`

Also not related to this PR. It just needs that variable to be removed as it's not used.

escopecz commented:
Another issue. When I try to create a new form action "Send email to contact"

`PHP Warning - array_diff(): Argument #1 is not an array in app/bundles/CoreBundle/Form/ChoiceLoader/EntityLookupChoiceLoader.php:148`

dongilbert commented:
Thanks for the reports @escopecz - looking into fixing them now

escopecz commented:
The last error is already fixed in https://github.com/mautic/mautic/pull/8276/files#diff-aee07ed1ecc5d7efb0d5f3fca2f73775R30, so this is ready for merge
